### PR TITLE
Improve ride rating calculation

### PR DESCRIPTION
**Pull request by ZehMatt**
A couple of things this PR does:
- Has now 4 states that can be updated rather than a single ride only.
- The previous implementation did not skip shops/facilities and instead spent an entire tick to reset the state machine, now it filters rides beforehand so it wont potentially do one tick nothing but initialize data and discard it the next tick. it also considers some other ride states.
- Updating the ride state now has sub-stepping updates to process more data without rewriting the entire state machine, because there is still a problem with hacked rides we can't allow processing it all at once currently, there seems to be no real performance impact doing it like this, I did a bit of testing and 20 sub steps seem to hardly impact the overall framerate, could probably raise this even further but it now updates all rides in a couple of seconds which seemed good enough to me.

My testing shows that iterating and updating all rides takes only a couple seconds now. This will probably break bpb replay as the calculation goes a lot faster now.

### CODE REVIEW

**Comment by ocalhoun6** [comment]
I'm concerned that this part might be a problem:
`    // Skip rides that are closed.
    if (ride.status == RideStatus::Closed)
    {
        return true;
    }`

Often, when I'm first testing a new coaster, I'll put it in test mode just long enough for the first train to depart, and then put it back into closed mode for the duration of the test. (This makes it simpler to watch, avoiding crowding the track with trains, and it can help avoid crashes if a train valleys out.)

Currently, this works fine and still generates ride ratings once the train completes the circuit. But if the ride rating calculator doesn't look at closed rides, maybe this would no longer work, and now the ride won't be able to get ratings until it completes a full circuit while in open or testing mode?

**Comment by ZehMatt** [comment]
> I'm concerned that this part might be a problem:
> 
> ```
> // Skip rides that are closed.
> 
>     if (ride.status == RideStatus::Closed)
> 
>     {
> 
>         return true;
> 
>     }
> ```
> 
> Often, when I'm first testing a new coaster, I'll put it in test mode just long enough for the first train to depart, and then put it back into closed mode for the duration of the test. (This makes it simpler to watch, avoiding crowding the track with trains, and it can help avoid crashes if a train valleys out.)
> 
> Currently, this works fine and still generates ride ratings once the train completes the circuit. But if the ride rating calculator doesn't look at closed rides, maybe this would no longer work, and now the ride won't be able to get ratings until it completes a full circuit while in open or testing mode?

This hasn't changed and you need to be lucky to open/close the right at the time it picked up the ride calculation also, that condition already existed.

File: src/openrct2/ride/RideRatings.cpp [hunk 1]
```diff
@@ -6,7 +6,6 @@
  *
  * OpenRCT2 is licensed under the GNU General Public License version 3.
  *****************************************************************************/
-
```
Comment (Broxzier) [inline src/openrct2/ride/RideRatings.cpp:9]: Accidental line change

File: src/openrct2/ride/RideRatings.cpp [hunk 2]
```diff
@@ -90,9 +94,27 @@ static void ride_ratings_begin_proximity_loop(RideRatingUpdateState& state);
...
+
+void RideRatingResetUpdateStates()
+{
+    RideRatingUpdateState nullState{};
+    nullState.State = RIDE_RATINGS_STATE_FIND_NEXT_RIDE;
+
+    std::fill(gRideRatingUpdateStates.begin(), gRideRatingUpdateStates.end(), nullState);
+}
+
+RideRatingUpdateState& RideRatingGetUpdateState(size_t index)
```
Comment (Broxzier) [inline src/openrct2/ride/RideRatings.cpp:112]:
```suggestion
static RideRatingUpdateState& RideRatingGetUpdateState(size_t index)
```

**Review by Broxzier — CHANGES_REQUESTED** [review]
Code changes LGTM. A few minor points. When testing the build, I noticed that the Excitement/Intensity/Nausea rating appeared instantly. Is this intended?

File: src/openrct2/park/ParkFile.cpp [hunk 3]
```diff
@@ -536,7 +536,25 @@ namespace OpenRCT2
                 cs.ReadWrite(gGrassSceneryTileLoopPosition);
                 cs.ReadWrite(gWidePathTileLoopPosition);
 
-                ReadWriteRideRatingCalculationData(cs, gRideRatingUpdateState);
+                auto& rideRatings = RideRatingGetUpdateStates();
+                if (os.GetHeader().TargetVersion >= 21)
+                {
+                    cs.ReadWriteArray(rideRatings, [this, &cs](RideRatingUpdateState& calcData) {
```
[the remark this answers is not in the document]
Comment (ZehMatt) [reply src/openrct2/park/ParkFile.cpp:542]: No, it wont be able to call members without this.

**Comment by ZehMatt** [comment]
> Code changes LGTM. A few minor points. When testing the build, I noticed that the Excitement/Intensity/Nausea rating appeared instantly. Is this intended?

If its correct then yes, its supposed should be a lot quicker

**Comment by Broxzier** [comment]
Normally when a vehicle finishes a lap, the statistics appear, but the ratings appeared much later. I thought this was intentional, desired behaviour. That the ratings appear much sooner now seems like a good change to me, I was just wondering if this was an intended change.

**Comment by ZehMatt** [comment]
> Normally when a vehicle finishes a lap, the statistics appear, but the ratings appeared much later. I thought this was intentional, desired behaviour. That the ratings appear much sooner now seems like a good change to me, I was just wondering if this was an intended change.

Yeah they were massively delayed because of a lot of factors, each tick would only do a few calculations for a single ride, it also spent ticks doing nothing if the ride id it encountered has been a shop, it would basically enter the next state in the state machine and on the next tick determines it has no stations and resets the state back to "find next ride", I skip all rides now where the required condition fails to process it. Its still not exactly instant but it should only take a couple seconds now if you are unlucky and the current tick processed the last ride where it then starts over with the first ride.

File: src/openrct2/ride/RideRatings.cpp (hunk 2 above)
Comment (ZehMatt) [reply src/openrct2/ride/RideRatings.cpp:112]: I actually forgot about this, it's no longer used, this was to keep each commit in a working state, ill remove it.

File: src/openrct2/park/ParkFile.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Broxzier) [reply src/openrct2/park/ParkFile.cpp:542]: Right, for some reason I thought the function call in the lambda was a free function. This makes sense now.

File: src/openrct2/park/ParkFile.h [hunk 4]
```diff
@@ -9,7 +9,7 @@ struct ObjectRepositoryItem;
 namespace OpenRCT2
 {
     // Current version that is saved.
-    constexpr uint32_t PARK_FILE_CURRENT_VERSION = 20;
+    constexpr uint32_t PARK_FILE_CURRENT_VERSION = 21;
 
     // The minimum version that is forwards compatible with the current version.
     constexpr uint32_t PARK_FILE_MIN_VERSION = 19;
```
Comment (duncanspumpkin) [inline src/openrct2/park/ParkFile.h:15]: you'll need to bump the min version as the new save can't be read by version 20.

**Review by Broxzier — APPROVED** [review]

**Comment by cheweytoo** [comment]
Just for the record: This. Is. Amazing! I consider this to be one of the greatest QoL improvements in OpenRCT2 ever.

Thank you!

**Pull request merged**